src/module.c

```diff
@@ -12244,6 +12244,134 @@ int RM_LoadConfigs(RedisModuleCtx *ctx) {
     return REDISMODULE_OK;
 }
 
+/* --------------------------------------------------------------------------
+ * ## RDB load/save API
+ * -------------------------------------------------------------------------- */
+
+#define REDISMODULE_RDB_STREAM_FILE 1
+
+typedef struct RedisModuleRdbStream {
+    int type;
+
+    union {
+        char *filename;
+    } data;
+} RedisModuleRdbStream;
+
+/* Create a stream object to save/load RDB to/from a file.
+ *
+ * This function returns a pointer to RedisModuleRdbStream which is owned
+ * by the caller. It requires a call to RM_RdbStreamFree() to free
+ * the object. */
+RedisModuleRdbStream *RM_RdbStreamCreateFromFile(const char *filename) {
+    RedisModuleRdbStream *stream = zmalloc(sizeof(*stream));
+    stream->type = REDISMODULE_RDB_STREAM_FILE;
+    stream->data.filename = zstrdup(filename);
+    return stream;
+}
+
+/* Release an RDB stream object. */
+void RM_RdbStreamFree(RedisModuleRdbStream *stream) {
+    switch (stream->type) {
+    case REDISMODULE_RDB_STREAM_FILE:
+        zfree(stream->data.filename);
+        break;
+    default:
+        serverAssert(0);
+        break;
+    }
+    zfree(stream);
+}
+
+/* Load RDB file from the `stream`. Dataset will be cleared first and then RDB
+ * file will be loaded.
+ *
+ * `flags` must be zero. This parameter is for future use.
+ *
+ * On success REDISMODULE_OK is returned, otherwise
+ * REDISMODULE_ERR is returned and errno is set to the following values:
+ *
+ * * EINVAL: `stream` is NULL or `flags` value is invalid.
+ * * EIO: Failed to load the RDB file. Check server logs for more info.
+ *
+ * Example:
+ *
+ *     RedisModuleRdbStream *s = RedisModule_RdbStreamCreateFromFile("exp.rdb");
+ *     RedisModule_RdbLoad(s, 0);
+ *     RedisModule_RdbStreamFree(s);
+ */
+int RM_RdbLoad(RedisModuleRdbStream *stream, int flags) {
+    if (!stream || flags != 0) {
+        errno = EINVAL;
+        return REDISMODULE_ERR;
+    }
+
+    if (server.aof_state != AOF_OFF) stopAppendOnly();
+
+    /* Kill existing RDB fork as it is saving outdated data. Also killing it
+     * will prevent COW memory issue. */
+    if (server.child_type == CHILD_TYPE_RDB) killRDBChild();
+
+    emptyData(-1,EMPTYDB_NO_FLAGS,NULL);
+
+    /* rdbLoad() can go back to the networking and process network events. If
+     * RM_RdbLoad() is called inside a command callback, we don't want to
+     * process the current client. Otherwise, we may free the client or try to
+     * process next message while we are already in the command callback. */
+    if (server.current_client) protectClient(server.current_client);
+
+    serverAssert(stream->type == REDISMODULE_RDB_STREAM_FILE);
+    int ret = rdbLoad(stream->data.filename,NULL,RDBFLAGS_NONE);
```

Comment:
good catch, i agree it should be allowed only on masters, and it should drop replicas when called.

Reply:
I'd also like us to evaluate if this API is really needed or if there is an alternative strategy here for loading. I assume this being used to boot strap, or re-provision an existing node, to be added into the cluster, so perhaps the better way is to support this like normal node bootsrapping that the module can tap into. The specific use case trying to be solved here isn't super clear.

Reply:
For our use case, a node can load rdb anytime not just for bootstrapping and it can happen multiple times.
I'm not that familiar with master-replica setup, so if you think we should block it on replicas, I can make the required changes.

Reply:
Please be specific when you say an RDB can be loaded "anytime". AFAIK, this is for RedisRaft, which makes me believe that you are doing it when nodes are being added to the raft group or when you've fallen out of sync and want to restore it from the snapshot of another node. That is not "anytime", but is bootstrapping a new node into the cluster, the same way we do for replication.

An alternative is could configure Redis into the normal restoring state, so that we could arbitrarily load a snapshot from some location specified by the module. This would also enable other use cases like doing an online restore from a snapshot in S3 or something.

Reply:
Okay, sorry it wasn't clear. You are correct about our use case. I think we are talking about the same thing but using different terms. I thought you were using "bootstrapping" as if like something can be done only once for a Redis process. 

As Redis doesn't know about my clustering implementation, from Redis point of view, I can call `RM_RdbLoad()` anytime:
- I can start Redis, before running any command, I can call `RM_RdbLoad()`
- I can run some commands and change keyspace, then call `RM_RdbLoad()` . I can even call `RM_RdbLoad()` multiple times over time, I can run some other commands between these `RM_RdbLoad()` calls etc.

So, maybe that is bootstrapping in Redis, I'm not familiar with replication code.   

Regarding second part of your comment, with current `RM_RdbLoad()` API, wouldn't it work for who wants to load RDB from S3 or somewhere? You can write a module to download it and call `RM_RdbLoad()`?

Are you saying that we can put code in Redis to pull RDB from a remote destination? If so, for redisraft I think this is something I want to handle in the module. Because each chunk we download has some metadata and we are doing error handling (we may want to abort downloading RDB or indicate failure by replying to the sender etc)  Also, chunks (and their replies) indicate liveness of the nodes. We also control delivery speed as well and want to adjust it depending on the cluster situation. Pushing all (probably not all but some) of these into redis and handling other parts with workarounds in the module would be too complex for us unfortunately.

Reply:
I pushed a commit to drop replicas on master and to prevent RM_RdbLoad() on replicas, not sure if I got it correct though. Please take a look.

Reply:
option [1] may have other undesired complications (like a PID change, disconnecting some sockets, etc).

option [2] is kinda where i'm aiming for (mimic what happens when a replica does a full sync).
but if i understand correctly, you're suggesting to leave most of redis's code unchanged and trick it into thinking it's getting an RDB file from a master, which would be an awkward piece of code (i suppose in redis as well), and could have some complications that are hard to work around.
instead I think we can add the code to properly support it (without any user facing command yet), and let modules (which have narrower list of use cases and concerns) use it and get it matured.

Reply:
i think that suggestion [2] is to simulate something like what happens after a REPLICAOF command, i.e. a module API that initiate replication from a file descriptor (that's actually a pipe), and then redis will do the rest. as i said, i don't think i like this approach.